narwhals/_pyspark/dataframe.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+from typing import Any
+from typing import Iterable
+from typing import Sequence
+
+from narwhals._pyspark.utils import parse_exprs_and_named_exprs
+from narwhals._pyspark.utils import translate_sql_api_dtype
+from narwhals.utils import Implementation
+from narwhals.utils import flatten
+from narwhals.utils import parse_columns_to_drop
+from narwhals.utils import parse_version
+
+if TYPE_CHECKING:
+    from pyspark.sql import DataFrame
+    from typing_extensions import Self
+
+    from narwhals._pyspark.expr import PySparkExpr
+    from narwhals._pyspark.group_by import PySparkLazyGroupBy
+    from narwhals._pyspark.namespace import PySparkNamespace
+    from narwhals._pyspark.typing import IntoPySparkExpr
+    from narwhals.dtypes import DType
+    from narwhals.typing import DTypes
+
+
+class PySparkLazyFrame:
+    def __init__(
+        self,
+        native_dataframe: DataFrame,
+        *,
+        backend_version: tuple[int, ...],
+        dtypes: DTypes,
+    ) -> None:
+        self._native_frame = native_dataframe
+        self._backend_version = backend_version
+        self._implementation = Implementation.PYSPARK
+        self._dtypes = dtypes
+
+    def __native_namespace__(self) -> Any:  # pragma: no cover
+        if self._implementation is Implementation.PYSPARK:
+            return self._implementation.to_native_namespace()
+
+        msg = f"Expected pyspark, got: {type(self._implementation)}"  # pragma: no cover
+        raise AssertionError(msg)
+
+    def __narwhals_namespace__(self) -> PySparkNamespace:
+        from narwhals._pyspark.namespace import PySparkNamespace
+
+        return PySparkNamespace(
+            backend_version=self._backend_version, dtypes=self._dtypes
+        )
+
+    def __narwhals_lazyframe__(self) -> Self:
+        return self
+
+    def _from_native_frame(self, df: DataFrame) -> Self:
+        return self.__class__(
+            df, backend_version=self._backend_version, dtypes=self._dtypes
+        )
+
+    @property
+    def columns(self) -> list[str]:
+        return self._native_frame.columns  # type: ignore[no-any-return]
+
+    def collect(self) -> Any:
+        import pandas as pd  # ignore-banned-import()
+
+        from narwhals._pandas_like.dataframe import PandasLikeDataFrame
+
+        return PandasLikeDataFrame(
+            native_dataframe=self._native_frame.toPandas(),
+            implementation=Implementation.PANDAS,
+            backend_version=parse_version(pd.__version__),
+            dtypes=self._dtypes,
+        )
```

Comment:
`pandas` (as well as `pyarrow` and `numpy`) is still a dependency of PySpark https://github.com/apache/spark/blob/master/python/packaging/classic/setup.py#L350-L359

`toPandas()` will not fail, but I agree that it would be nice to give the user a choice.

To me, this sounds like a followup what do you think?